### PR TITLE
Fixed many issues in grapheme walker

### PR DESCRIPTION
**Pull request by dukc**
I had the Unicode table generator to use a few more tables from the Unicode database, correcting many issues. For example, being a control code from grapheme walking perspective is not the same as the code point belonging to the Cc general category, unlike how our implementation has done it. This fixes [Issue 23474](https://issues.dlang.org/show_bug.cgi?id=23474) among other unreported issues.

I also had the implementation to consider Emoji sequences and prefixed extension characters, and added tests for those.

Unfortunately, I currently have issues with 32-bit DMD so I could not regenerate the Unicode tables for this PR. This is why it's still draft. I should be able to fix that, but it might be quicker if someone else regenerates the Unicode tables with the generator modifications here - anyone?

See also: [Grapheme iteration in the Unicode Standard](https://www.unicode.org/reports/tr29/tr29-41.html#Default_Grapheme_Cluster_Table)

### CODE REVIEW

**Comment by rikkimax** [comment]
> Unfortunately, I currently have issues with 32-bit DMD so I could not regenerate the Unicode tables for this PR.

Yeah that is a pretty bad problem we've got there, that I'm not happy with.

Because it ties into the WORD size, I think the solution may end up involving splitting std.uni and allowing you to pass in whatever type you want. But I'm not 100% certain.

**Comment by dukc** [comment]
16 unrelated commits added. Bad Git usage from me. Ah well, rebasing should fix it.

**Comment by dukc** [comment]
Why does Github think I'm trying to recommit everything that happened after my original commit but before Rikki's changes? If I run `git log master..HEAD`, only the five commits of me and Rikki are listed. I also did `git rebase master --rebase-merges`, then `git push`, but it says everything is up to date.

**Comment by SixthDot** [comment]
use the action "drop" for the unrelated commits when you do your interactive rebase.

**Comment by dukc** [comment]
The interactive rebase lists only the five commits I want to keep, to begin with.

**Comment by SixthDot** [comment]
maybe try `git checkout master && git pull` then `git checkout correctGraphemeWalker && git rebase -i master` ?

**Comment by dukc** [comment]
Already done, with the difference that `git rebase -i` only works correctly with `--rebase-merges`, because first of the five commits is older than the master branch with the second commit merging them together among other changes.

**Comment by SixthDot** [comment]
okay, sorry for my useless comments and good luck with that.

**Comment by dukc** [comment]
Thanks for trying anyway.

**Comment by rikkimax** [comment]
The only solution I know: reset to before changes. Pull from upstream. Copy
my files over. Force push.

Easy :)

**Comment by dukc** [comment]
Finally! I rebased with `--rebase-merges`, but applying my last-year commit on top of master head. That resulted in some strange rebase conflicts and the last commit being reattributed to me, but I think it works now. Still going to improve readability of `GenericDecodeGrapheme` so don't merge yet.

**Comment by dukc** [comment]
@rikkimax Here you go. I totally refactored the whole function. I'm afraid this might hider performance though. By no means I except this to be slow, but function calls just aren't as fast as gotos.

**Comment by rikkimax** [comment]
``GraphemeState``, ``TransformRes``, ``graphemeTransforms`` can all go under ``genericDecodeGrapheme`` template. They won't need to be private (they should already be).

The one thing I want to see is the TR29 rules listed to which code.  Like this: https://github.com/Project-Sidero/basic_memory/blob/main/source/sidero/base/text/unicode/wordbreak.d#L87
That'll make it easy to scan which rules are currently implemented and confirm that they currently match the spec.

Otherwise, this is looking quite good. I need to remember that you can specify array elements like that because that is a neat trick with graphemeTransforms!

**Comment by RazvanN7** [comment]
ping @dukc

**Comment by dukc** [comment]
> ping @dukc

Yes, I got the note above. I'm going to add the commentary as Rikki requested, it's just that I tend to be too lazy to get work done other than on Saturdays and holidays.

**Comment by dukc** [comment]
Nonetheless, done now.

File: std/uni/package.d [hunk 1]
```diff
@@ -6962,139 +6964,235 @@ private:
...
+//   or extend | '\u200D' separately
 
 private static bool isRegionalIndicator(dchar ch) @safe pure @nogc nothrow
 {
     return ch >= '\U0001F1E6' && ch <= '\U0001F1FF';
 }
 
+// Our grapheme decoder is a state machine, this is list of all possible
+// states before each code point.
+private enum GraphemeState {
```
Comment (dkorpel) [inline std/uni/package.d:6978]:
```suggestion
private enum GraphemeState 
{
```

File: std/uni/package.d [hunk 2]
```diff
@@ -6962,139 +6964,235 @@ private:
...
+    V,
+    LVT,
+    Emoji,
+    EmojiZWJ,
+    Prepend,
+    End
+}
+
+// Message values whether end of grapheme is reached
+private enum TransformRes {
```
Comment (dkorpel) [inline std/uni/package.d:6992]:
```suggestion
private enum TransformRes 
{
```

**Comment by dukc** [comment]
@dkorpel Thanks, fixed. It seems the brace style in Symmetry project is becoming a habit for me.

Also added a changelog entry.

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**